# Settings

In [1]:
import pandas as pd
import numpy as np
import math

from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.preprocessing import MinMaxScaler

from matplotlib import pyplot
%matplotlib inline

# Read dataset

In [2]:
df = pd.read_csv('../RNN_data_version1.csv')
df = df.drop(df.columns[0], axis=1)

In [5]:
df.head(200)

activity1   builtin1  communication1  entertainment1  finance1     game1  \
0     0.000000      0.000           0.000           0.000     0.000     0.000   
1     0.000000      0.000           0.000           0.000     0.000     0.000   
2     0.000000      0.000           0.000           0.000     0.000     0.000   
3     0.000000      0.000           0.000           0.000     0.000     0.000   
4     0.081548    248.979        2168.229         350.856     0.000     0.000   
5     0.134050   3139.218        6280.890        1007.456    49.544     0.000   
6     0.236880    731.429        4962.918          93.324    21.076     0.000   
7     0.142741   1286.246        5237.319          94.346    43.403     0.000   
8     0.078961    866.956        9270.629         976.971    34.106     0.000   
9     0.098374   1032.768       10276.751          68.206    43.054     0.000   
10    0.101308   1167.497        8988.753         910.479    52.331     0.000   
11    0.159511   1229.327        6936.512         639.339    42.219     0.000   
12    0.095698  10062.595         866.833        1005.403    89.166   233.036   
13    0.068203   1952.630        2720.692        1627.429     0.000     0.000   
14    0.049093    414.365        1298.505        1535.677     0.000     0.000   
15    0.070505   2961.405        5332.637        1758.920    41.255    51.176   
16    0.118443   1385.517        6095.526        1030.637   547.826   234.741   
17    0.101917   3614.349        6920.156        1343.376   247.777     0.000   
18    0.049238   2321.464       14438.082        1293.631    76.234     0.000   
19    0.126689   1957.829        5481.628        1545.227     0.000     0.000   
20    0.119002  15159.190        1885.409          76.369     0.000     0.000   
21    0.064408   4548.662        3355.743         187.214     0.000   114.406   
22    0.050671    927.377        7355.015         447.121    46.128     0.000   
23    0.075327    892.689        9477.553        3466.479    19.131     0.000   
24    0.120384   1638.617        6177.076         940.770     0.000     0.000   
25    0.078963    715.355        4881.411        1858.893     0.000     1.059   
26    0.079986   3369.031        3336.592        1976.138    94.093     0.000   
27    0.072249    302.446        6595.083        2186.482     0.000     0.000   
28    0.014021    447.415        1096.795         249.392    82.413     0.000   
29    0.083887    961.581        6486.284        6265.556    89.323     0.000   
..         ...        ...             ...             ...       ...       ...   
170   0.101777    433.548        5939.196         366.749     0.000     0.000   
171   0.062848    525.800        2224.006         279.441     0.000     0.000   
172   0.080947    244.856        3439.813         966.003     0.000     0.000   
173   0.039371    440.236        3924.221          99.176     0.000     0.000   
174   0.049053   1591.018        3506.227         838.425     0.000     0.000   
175   0.200166    688.974        3364.003        1170.485     0.000     0.000   
176   0.018822    130.072        3169.602         230.117     0.000     0.000   
177   0.010305    427.954        4711.910         671.527     0.000     0.000   
178   0.041165      0.000        2739.915         583.655     0.000     0.000   
179   0.000000      0.000           0.000           0.000     0.000     0.000   
180   0.000000      0.000           0.000           0.000     0.000     0.000   
181   0.000000      0.000           0.000           0.000     0.000     0.000   
182   0.000000      0.000           0.000           0.000     0.000     0.000   
183   0.043432     43.495        1100.514        1073.377     0.000     0.000   
184   0.137709    116.887        3930.897        2999.969     0.000     0.000   
185   0.135563    212.512        3234.500        1950.957     0.000     0.000   
186   0.132329    523.696        4045.840        2453.259     0.000     0.000   
187   0.171471     42.311        358

In [3]:
df.shape

(1181, 96)

# Make data ready for input

This data has 1181 samples with for every sample 5 timesteps. Every timestep has 19 different numerical features.

Use normalisation of range (0,1) to counter for vanishing gradient problem and other advantages.

In [5]:
# Split into train and test set
dataset = df.iloc[:,:].values
rand_split = np.random.rand(len(dataset))
train_list = rand_split < 0.8
#val_list = (rand_split >= 0.8) & (rand_split < 0.9)
test_list = rand_split >= 0.8

train = dataset[train_list]

# normalize features
scaler = MinMaxScaler()
train = scaler.fit_transform(train)

test = dataset[test_list]

# split into input and output
train_X, train_y = train[:,:-1], train[:,-1]
test_X, test_y = test[:,:-1], test[:,-1]

# reshape input to be 3D [samples, timesteps, features]
train_X = train_X.reshape((train_X.shape[0], 5, 19))
test_X = test_X.reshape((test_X.shape[0], 5, 19))
print(train_X.shape, train_y.shape, test_X.shape, test_y.shape)

(933, 5, 19) (933,) (248, 5, 19) (248,)


In [22]:
np.square(scaler.data_max_[95] - scaler.data_min_[95])

36.0

# Train model

Many-to-one LSTM recurrent neural network with 32 neurons, dropout of 20% for regularization and adam optimizer.

### Long Short-Term Memory (LSTM) network

This following figure shows the operations of an LSTM-cell.

<img src="images/lstm.png" style="width:700px;height:300px;">

### Nested k-fold cross-validation

This is done to evaluate between different models.

In [8]:
# function to create model for kerasClassifier
def create_model(dropout_rate, neurons):
    from keras.models import Sequential
    from keras.layers import Reshape
    from keras.layers import Dense
    from keras.layers import LSTM
    from keras.layers import Dropout
    model = Sequential()
    #model.add(Reshape((5,19), input_shape=(95,)))
    model.add(LSTM(units=neurons, input_shape=(5, 19), dropout=dropout_rate))
    model.add(Dense(1))
    model.compile(loss='mse', optimizer='adam', metrics=['mse'])
    return model

# create model
def train_model():
    from keras.wrappers.scikit_learn import KerasRegressor
    model = KerasRegressor(build_fn=create_model, verbose=0)
    
    # define the grid search parameters
    batch_size = [64]
    epochs = [128, 256]
    dropout_rate = [0.0, 0.2]
    neurons = [32,64,128]
    param_grid = dict(batch_size=batch_size, epochs=epochs, dropout_rate=dropout_rate, neurons=neurons)
    grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=1)
    grid_result = cross_val_score(grid, train_X, train_y, cv=3, verbose=1)
    return grid_result
    
results = train_model()



[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed: 43.8min finished


In [47]:
metrics = (np.absolute(results) * np.square(scaler.data_max_[95] - scaler.data_min_[95]))
print(np.mean(metrics))
print(metrics)

0.536661780089
[ 0.70661271  0.37828904  0.5250836 ]


In [6]:
pyplot.boxplot(metrics)

NameError: name 'metrics' is not defined

### K-fold cross-validation

Hyperparameter grid search tuning

In [ ]:
# function to create model for kerasClassifier
def create_model(dropout_rate, neurons):
    from keras.models import Sequential
    from keras.layers import Reshape
    from keras.layers import Dense
    from keras.layers import LSTM
    from keras.layers import Dropout
    model = Sequential()
    #model.add(Reshape((5,19), input_shape=(95,)))
    model.add(LSTM(units=neurons, input_shape=(5, 19), dropout=dropout_rate))
    model.add(Dense(1))
    model.compile(loss='mse', optimizer='adam', metrics=['mse'])
    return model

# create model
def train_model():
    from keras.wrappers.scikit_learn import KerasRegressor
    model = KerasRegressor(build_fn=create_model, verbose=0)
    
    # define the grid search parameters
    batch_size = [64]
    epochs = [128, 256]
    dropout_rate = [0.0, 0.2]
    neurons = [32, 64, 128]
    param_grid = dict(batch_size=batch_size, epochs=epochs, dropout_rate=dropout_rate, neurons=neurons)
    grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-1)
    grid_result = grid.fit(train_X, train_y)
    return grid_result
    
results = train_model()


In [ ]:
print("Best: %f using %s" % (results.best_score_, results.best_params_))
means = results.cv_results_['mean_test_score']
stds = results.cv_results_['std_test_score']
params = results.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
	print("%f (%f) with: %r" % (mean, stdev, param))

In [ ]:
# design neural net
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers import Dropout
model = Sequential()
model.add(LSTM(128, input_shape=(5, 19), dropout=0.0))
model.add(Dense(1))
model.compile(loss='mse', optimizer='adam')

# fit neural net
history = model.fit(train_X, train_y, epochs=128, batch_size = 64, validation_split=0.2, verbose = 2, shuffle=True)

In [ ]:
pyplot.plot(history.history['loss'], label='train')
pyplot.plot(history.history['val_loss'], label='validation')
pyplot.legend()
pyplot.show()

# Calculate and compare model accuracy

In [32]:
MAE = lambda p,o: sum(abs(p - o)) / len(o)
RMSE = lambda p,o: (sum((p - o) ** 2) / len(o))

In [30]:
originals = df.iloc[:,92].values
predictions = df.iloc[:,95].values

## Baseline accuracy

Baseline accuracy is calculated by predicting mood of t+1 with mood of time t. Both MAE and RMSE are showed.

In [34]:
print('MAE: {}'.format(MAE(predictions, originals)))
print('MSE: {}'.format(RMSE(predictions, originals)))

MAE: 0.550179943390397
MSE: 0.5670021934142934


Check how to statistically define a confidence interval for RMSE baseline value. Two approaches can be considered.

- Use assumptions that predictions - originals are normal distributed

\begin{equation}
H← ​​​60 ​+​ \frac{​​30(B−R)​​}{Vmax−Vmin}  ​​, if V​max​​ = G
\end{equation}

- Use bootstrap to estimate confidence interval of RMSE

In [ ]:
x = (predictions - originals)
np.mean(x)

In [ ]:
import scipy.stats as stats
stats.probplot(x, dist="norm", plot=pyplot)
pyplot.show()

In [ ]:
pyplot.hist(x)

## Model accuracy

Prediction of values with model for test set and then scale values back.

In [ ]:
#make predictions
trainPredict = model.predict(train_X)
testPredict = model.predict(test_X)

inverse = np.concatenate((test_X.reshape((test_X.shape[0], 95)), testPredict), axis=1)
inverse2 = scaler.inverse_transform(inverse)
testPrediction = inverse2[:,-1]

inverse = np.concatenate((test_X.reshape((test_X.shape[0], 95)), test_y.reshape((len(test_y), 1))), axis=1)
inverse2 = scaler.inverse_transform(inverse)
testY = inverse2[:,-1]

#invert predictions
#trainPredict = scaler.inverse_transform(trainPredict)
#trainY = scaler.inverse_transform(train_y)
#testPredict = scaler.inverse_transform(testPredict)
#testY = scaler.inverse_transform(test_y)
original = test_y[np.newaxis].T

In [ ]:
print('MAE: {}'.format(MAE(testPrediction, testY)))
print('MSE: {}'.format(RMSE(testPrediction, testY)))